redsky/event_streams.py

```diff
@@ -0,0 +1,711 @@
+
+    def stop(self, docs):
+        """
+        Issue new descriptor document for input documents
+
+        Parameters
+        ----------
+        docs: tuple of dicts or dict
+
+        Returns
+        -------
+        name: 'descriptor'
+        doc: dict
+            The document
+        """
+        if not self.event_failed:
+            if self.run_start_uid is None:
+                raise RuntimeError("Received RunStop before RunStart.")
+            new_stop = dict(uid=str(uuid.uuid4()),
+                            time=time.time(),
+                            run_start=self.run_start_uid)
+            if isinstance(docs, Exception):
+                self.event_failed = True
+                new_stop.update(reason=repr(docs),
+                                trace=traceback.format_exc(),
+                                exit_status='failure')
+            if not self.event_failed:
+                new_stop.update(exit_status='success')
+            self.outbound_descriptor_uid = None
+            self.run_start_uid = None
+            return 'stop', new_stop
+
+    def event_guts(self, docs):
+        """
+        Provide some of the event data as a dict, which may be used as kwargs
+
+        Parameters
+        ----------
+        docs: tuple of dicts
+
+        Returns
+        -------
+
+        """
+        return {input_kwarg: docs[position]['data'][data_key] for
+                input_kwarg, (data_key, position) in self.input_info.items()}
+
+    def issue_event(self, outputs):
+        """Issue a new event
+
+        Parameters
+        ----------
+        outputs: tuple, dict, or other
+            Data returned from some external functon
+
+        Returns
+        -------
+        new_event: dict
+            The new event
+
+        Notes
+        -----
+        If the outputs of the function is an exception no event will be
+        created, but a stop document will be issued.
+        """
+        if not self.event_failed:
+            if self.run_start_uid is None:
+                raise RuntimeError("Received Event before RunStart.")
+            if isinstance(outputs, Exception):
+                return self.stop(outputs)
+
+            # Make a new event with no data
+            if len(self.output_info) == 1:
+                outputs = (outputs,)
+
+            new_event = dict(uid=str(uuid.uuid4()),
+                             time=time.time(),
+                             timestamps={},
+                             descriptor=self.outbound_descriptor_uid,
+                             filled={k[0]: True for k in self.output_info},
+                             seq_num=self.i)
+
+            if self.output_info:
+                new_event.update(data={output_name: output
+                                       for (output_name, desc), output in
+                                       zzip(self.output_info, outputs)})
+            else:
+                new_event.update(data=outputs['data'])
```

Comment:
`output_info` is designed to do two things:
1. create a map between function returns and data keys
1. Provide information for building a descriptor for that data

Not all stream functions produce new data (eg filter) so they don't need new descriptors or a map between function returns and data keys.

In this case the `else` is there in case we need to issue a new event (with new uids, timestamps, etc.) but want all the data from the old event (`bundle` for example).